math/linear_recurrence_matrix.cpp

```diff
@@ -0,0 +1,370 @@
+/**
+ * @brief Evaluate recurrence relation using [matrix
+ * exponentiation](https://www.hackerearth.com/practice/notes/matrix-exponentiation-1/).
+ * @details
+ * Given a recurrence relation; evaluate the value of nth term.
+ * For e.g., For fibonacci series, recurrence series is `f(n) = f(n-1) + f(n-2)`
+ * where `f(0) = 0` and `f(1) = 1`.
+ * Note that the method used only demonstrates
+ * recurrence relation with one variable (n), unlike `nCr` problem, since it has
+ * two (n, r)
+ *
+ * ### Algorithm
+ * This problem can be solved using matrix exponentiation method.
+ * @see here for simple [number exponentiation
+ * algorithm](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/math/modular_exponentiation.cpp)
+ * or [explaination
+ * here](https://en.wikipedia.org/wiki/Exponentiation_by_squaring).
+ * @author [Ashish Daulatabad](https://github.com/AshishYUO)
+ */
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for std::vector STL
+
+/**
+ * @namespace math
+ * @brief Mathematical algorithms
+ */
+namespace math {
+/**
+ * @namespace linear_recurrence_matrix
+ * @brief Implementation of solution for linear recurrence using matrix
+ */
+namespace linear_recurrence_matrix {
+
+/**
+ * @brief Implementation of matrix multiplication
+ * @details Multiplies matrix A and B, given total columns in A are equal to
+ * total given rows in column B
+ * @tparam T template type for integer as well as floating values, default is
+ * long long int
+ * @param _mat_a first matrix of size n * m
+ * @param _mat_b second matrix of size m * k
+ * @returns `_mat_c` resultant matrix of size n * k
+ * Complexity: `O(n*m*k)`
+ * @note The complexity in this case will be O(n^3) due to the nature of the
+ * problem. We'll be multiplying the matrix with itself most of the time.
+ */
+template <typename T = int64_t>
```

Comment:
Wouldn't `template <size_t T>` be better here (same in other parts of the code)?

Reply:
The template is used when a user wants a preferred type (e.g., integer or floating values) when evaluating recurrence relation. It won't be anything related to the size of matrix. :)

Reply:
I'll update the other changes though.

Reply:
> when a user wants a preferred type (e.g., integer or floating values)

Ah, OK, thanks for the explanation. 🙂